### ISSUE TITLE
feat: Allow extension operations to have child graphs

### ISSUE THREAD

**Issue by ss2165** [CQCL/hugr#1546]
We originally did not allow this for simplicity, but there a number of applications where this would be very convenient. 

A motivating example: a "dagger" box that contains a unitary subgraph and means "evaluate the dagger of this graph". A rewrite may dagger the inner graph at compile time to eliminate the dagger node and inline.

While making the changes from #1433 seems to be a good time to consider this.

Relatedly, it would be beneficial to unify the encoding of all statically available hugrs (i.e. inside dfg, control flow nodes, function definitions, FunctionValue constants, and now also inside custom operations). 

Open questions include whether the child graph of a custom op is always a dataflow graph, or it can include other region types like cfg, module or the one inside`Conditional`.

**Comment by croyzor** [comment]
Could your example be done equally well with a const hugr wired into an opaque "Dagger" op?

**Comment by ss2165** [comment]
yes I think there are multiple possible encodings with the current expressivity but for _static_ subroutines I would say allowing children is the most natural. For example with your case the rewrites would fail if something earlier had manager to insert something between the constant and the op, assuming it was a runtime value.

**Comment by zrho** [comment]
I think this is an essential feature, as it opens up a lot of use cases (see https://github.com/CQCL/hugr/discussions/1155#discussioncomment-9678673).

After our conversation at confinuum, I've made the model data structure https://github.com/CQCL/hugr/pull/1542 allow for this. A node can have a sequence of regions, each of which is currently either a dataflow region or a control flow region. This works well with the builtins, and is converted appropriately between core and model:

 - The `cfg` operation expects a control flow region.
 - The `block` operation expects a data flow region.
 - The `dfg` operation expects a data flow region.
 - The `define-func` operation expects a data flow region.
 - The `tail-loop` operation expects a data flow region.
 - The `conditional` operation expects as many data flow regions as there are cases in the sum.
 
This design also addresses the issue we had with order: on the one hand, the nodes in a dataflow graph do not have any intrinsic order between them. On the other hand, `conditional` operations need ordered cases. Since the cases here are regions they are ordered, while nodes remain order-agnostic within their region.

Each region comes with source and target ports, which are the ports internal to the region. That removes the need for `input`, `output` and `exit` nodes (and potential `entry` nodes), therefore also removing reliance on node order within a region.

**Comment by ss2165** [comment]
> Currently we would model this by storing a Hugr inside the Dagger ExtensionOp

It's not clear to me how I would do this?

**Comment by ss2165** [comment]
That would be quite an ugly hack for what should be a well supported usage pattern. I'm happy to consider allowing const edges to extension ops to be the core implementation and keep the general child pattern to the model as Lukas outlines above.

I like the arguments about some degree of invisibility to the host graph being potentially a good thing. 

> we should make Ops embedding Hugrs more ergonomic
this is ultimately the only goal of the proposal, happy to tweak how we get there before beginning work

**Comment by zrho** [comment]
> it's adding complexity for no gain

I'm surprised by that, since to me it appears the exact opposite, lowering complexity while simultaneously giving massive gain.

Sure, some thought might have to be put into scoping rules. But by passing in some entirely separate Hugr, any custom operation taking a sub hugr that shouldn't be isolated (just like `tail-loop` or `cond` isn't) would have to be linked somehow. That's more complicated. Moreover availability of operations can also be controlled via the extension sets.

Similarly rewrites: A lot of incremental lowering can be performed by local rewrites. There's a bunch of lowering procedures that can be expressed with a box that is incrementally made smaller from the boundary by local rewrites. For instance this can be done for autodiff or incrementalisation. It's just like the process of computing a differential of an expression by hand: you add a `d/dx(...)` around your expression and then push it inwards step by step until you end up with a derivative-free expression again. But this requires rewrites that can operate across nesting boundaries. That is certainly possible with isolated hugr instances passed in as props, but that would be extremely complicated. 

This is exactly the feature that makes Ops embedding Hugrs more ergonomic. And if you want to treat a subgraph as opaque, you're free to never look at its subgraph.

> A type arg with a string which contains a serialised Hugr.

The only situation where this is not much more complex to deal with is when you never actually intend to look at the contained hugr. There is a reason why we do not do this for loops or cfgs.

**Comment by zrho** [comment]
> By constraining the hierarchy of the HUGR to a closed set of "parent" nodes, we make the constructed HUGR more useful, because we can know more about it.

It is more useful to you since you appear to go for a form of LLVM, but it becomes unwieldy to be point of being useless for applications that go beyond that. To me the idea of HUGR has been to be able to explore various kinds of exotic modes of computation under one roof. This is not just academic curiosity, but is concretely relevant for a research compiler in quantum computing. Striking the H out of HUGR also strikes out the U, since then the Oxford office will build its own IR that can't talk to ours since ours will be useless to them.

There is an awesome opportunity here: Quantinuum is the only place I know of that has an entire house filled with people that draw (hierarchical!) string diagrams for all kinds of types of computation; quantum, ML, quantum photonics, etc. So far that has been somewhat isolated from what we do in Cambridge, and to some extent isolated from practicality. HUGR is so close to being exactly the tool they would need to translate a lot of existing and novel research into actually running programs.

To me that's THE value proposition of HUGR and I struggle to understand its purpose without that.

**Comment by zrho** [comment]
Here’s a proposal that might bring some peace of mind: we could start with the requirement that all regions to custom operations are isolated wrt port connectivity. After declarative extensions, we can try to figure out a way to express ways in which that requirement can be loosened in a manageable way, with relatively low priority.

**Comment by ss2165** [comment]
Not really following the thread, just pointing to this other discussion as my currently preferred way of doing this in the core: https://github.com/CQCL/hugr/discussions/1342

**Comment by acl-cqc** [comment]
> Not really following the thread, just pointing to this other discussion as my currently preferred way of doing this in the core: [#1342](https://github.com/CQCL/hugr/discussions/1342)

Yes, if we do that (now an issue #1594), can we close this??

**Comment by ss2165** [comment]
Don't think so - I think this discussion is still unresolved as to whether refactoring that to work via one single graph is a good thing to do (but a lower priority discussion)

**Comment by acl-cqc** [comment]
This is in some ways an alternative to #3260 - it addresses the modifiers usecase thereof, but not the protocols-via-monomorphization.

One advantage I see here (vs having static references to FuncDefns) *is* the ability to support nonlocal edges. I feel the contract here is simple: the child subgraph uses a bunch of classical values (nonlocal edges cannot be linear); so those will have to be available for the op. Without knowledge of the parent op we can say no more, but that is sufficient** until compilation (at which point we will have to know about the parent op).

**sufficient but not optimal: of course we lose any ability to optimize *what value is passed in* to the child subgraph.

The alternative is turning nonlocal inputs into local ones (extra arguments to the statically-referenced FuncDefn), which is straightforward but annoying and makes extension ops "second class citizens" compared to TailLoop, Conditional, etc.

So looking at @doug-q's excellent questions:
> Should I apply this rewrite under this extension op?

A safe default is "no" ;), but if a rewrite is semantically preserving, then you could. (However, depending on what the extension op does, the performance effect of a rewrite might be different, e.g. it might pessimize rather than optimize).

> Which functions are called from my function? Should I include children of this extension op?

I would view function calls (even from within extension ops) as inlinable. So yes. (It may be that the extension op calls this multiple times, say. Or needs to call a daggered version of the callee - but in which case we need to preserve that call until we build the daggered version, which we should do early.)

> How many uses does this port have? It is used non-locally under an extension op.

One. The classical value must be sent into the extension op. It can be copied or discarded when it gets there.

> What extensions are allowed under this extension op? They are not necessarily related to the extensions available to the parent.
> When I'm lowering one extension to another, should that happen under this extension op?

These last two are really good questions, but note they also happen under the alternative formulation with static function inputs; indeed, they look even worse (because the subgraph that would be inside the extension op, with a unique parent, may have *multiple* static uses!). I don't think we have a good answer here other than that we want to do modifier resolution ASAP before optimization - which is another way of saying, optimization should be conservative (fallback to doing nothing, and raise an error if trying to lower an unknown parent op - that's probably better than current behaviour of applying lowering to the used FuncDefn and then moaning about the custom parent extension op later).

**Comment by doug-q** [comment]
With function types losing extension lists some objections go away.

I agree non-local edges do not seem to be a problem, and I think they are a real advantage: we have seen that they can reduce edge counts enough to affect performance.

I do think allowing child regions requires huge amounts of change. Every walk of the hugr will need new treatment. The alternative of allowing static function type args will flow through nicely everywhere.

I assume we would allow multiple child regions, so you'd have to specify on your op what the immediate child looks like, i.e. CFG has ExitBlock/DataflowBlock and Conditional has Case. You'd have to specify what the signatures of those blocks are.

There is utility in allowing the solution, either child regions oir static function args, to have polymorphic signatures. This lets you prevent values escaping (i.e. see ST monad https://hackage-content.haskell.org/package/base-4.22.0.0/docs/Control-Monad-ST.html#t:ST). At present no ops have child regions with polymorphic signatures.

We need a story for how dataflow analysis will work. I think it's much simpler with static functions. 

Our pass scoping technology is well suited to being pointed at static functions or excluding static functions

**Comment by acl-cqc** [comment]
> With function types losing extension lists some objections go away.
> 
> I agree non-local edges do not seem to be a problem, and I think they are a real advantage: we have seen that they can reduce edge counts enough to affect performance.
> 
> I do think allowing child regions requires huge amounts of change. Every walk of the hugr will need new treatment. The alternative of allowing static function type args will flow through nicely everywhere.

Yes, I think this is true. Child regions feels like more change, largely for the "win" of nonlocal edges. Although guaranteed "treeness" (as opposed to arbitrary-graph of static references between functions) could also be a win.

> I assume we would allow multiple child regions, so you'd have to specify on your op what the immediate child looks like, i.e. CFG has ExitBlock/DataflowBlock and Conditional has Case. You'd have to specify what the signatures of those blocks are.

I was thinking single dataflow region might be sufficient, but we could go further....

> There is utility in allowing the solution, either child regions oir static function args, to have polymorphic signatures. This lets you prevent values escaping (i.e. see ST monad https://hackage-content.haskell.org/package/base-4.22.0.0/docs/Control-Monad-ST.html#t:ST). At present no ops have child regions with polymorphic signatures.

Haven't grokked this. Values inside the child subgraph can't escape except through the Output node and then the op, which is op-defined...

> We need a story for how dataflow analysis will work. I think it's much simpler with static functions.

Much like when we see a static reference to a function that we don't understand, dataflow just has to conservatively assume that every input is "top". (We've already had this problem with global map/with, leading us to conservatively assume that if a function loaded by LoadFunction is used *anywhere except a CallIndirect*, all inputs inside the function must be Top)

> Our pass scoping technology is well suited to being pointed at static functions or excluding static functions

That's a good point! Indeed we have no mechanism for handling extension-op subgraphs. The scoping is hierarchy-based (not function-edge based) though so it might not be too hard to generalize.

I've seen another problem: there is no general solution for doing linearization into extension ops (there is no guaranteed recipe for what change to make to an op to add a value to the Input node inside it)

**Comment by acl-cqc** [comment]
Also maybe worth noting that this does not help with Quantinuum/guppylang#2355 (*if* we want that), whereas #3260 does.
